### PR TITLE
Implementing Kafka Protocol Binding for CloudEvents using rdkafka-rust

### PR DESCRIPTION
**Pull request by pranav-bhatt**
This PR adds the cloudevents-sdk-rdkafka library, which provides Kafka protocol bindings for cloudevents-rust. 

Things left to do:-
a) Add documentation
b) [Configuring the key attribute](https://github.com/cloudevents/spec/blob/v1.0/kafka-protocol-binding.md#31-key-attribute)
c) Fine tuning

### CODE REVIEW

File: cloudevents-sdk-rdkafka/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  if the other call ends early then the SDK call will start early as well. See
+  the
+  [CloudEvents meeting minutes](https://docs.google.com/document/d/1OVF68rpuPK5shIHILK9JOqlZBbfe91RNzQ7u_P7YCDE/edit#)
+  to determine which week will have the call.
+- Slack: #cloudeventssdk (or #cloudevents-sdk-rust) channel under
+  [CNCF's Slack workspace](https://slack.cncf.io/).
+- Email: https://lists.cncf.io/g/cncf-cloudevents-sdk
+- Contact for additional information: Francesco Guardiani (`@slinkydeveloper` on slack).
+
+[Crates badge]: https://img.shields.io/crates/v/cloudevents-sdk-reqwest.svg
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/README.md:25]: wrong links here

File: cloudevents-sdk-rdkafka/src/headers.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            (s, attribute_name_to_header!(s))
+        }
+    })
+    .collect()
+}
+
+lazy_static! {
+    pub(crate) static ref ATTRIBUTES_TO_HEADERS: HashMap<&'static str, String> =
+        attributes_to_headers(SpecVersion::all_attribute_names());
+    pub(crate) static ref SPEC_VERSION_HEADER: &'static str = "ce_specversion";
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/headers.rs:33]: for static strings you don't need the `lazy_static` macro, you can just use the static keyword https://doc.rust-lang.org/1.29.0/book/first-edition/const-and-static.html#static

Comment (pranav-bhatt) [reply cloudevents-sdk-rdkafka/src/headers.rs]: done!

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            (true, _) => Encoding::STRUCTURED,
+            (_, Some(_)) => Encoding::BINARY,
+            _ => Encoding::UNKNOWN,
+        }
+    }
+}
+
+/// Method to transform an incoming [`Response`] to [`Event`]
+pub fn owned_record_to_event(res: OwnedMessage) -> Result<Event> {
+    MessageDeserializer::into_event(ConsumerRecordDeserializer::owned_new(res))
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:138]: msg: OwnedMessage

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+pub fn borrowed_record_to_event(res: &BorrowedMessage) -> Result<Event> {
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:141]: same as above

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    Result, StructuredDeserializer, StructuredSerializer,
+};
+use cloudevents::{message, Event};
+use rdkafka::message::{BorrowedMessage, Headers, Message, OwnedMessage};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::str;
+
+pub struct ConsumerRecordDeserializer {
+    pub(crate) headers: HashMap<String, Bytes>,
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:14]: For the header values and the payload, more than copying the bytes into `Bytes`, can we store the `&[u8]`? With some lifetimes, that should be doable because `ConsumerRecordDeserializer` for sure lives less than both `OwnedMessage` and `&BorrowedMessage`

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+            .unwrap())?)?;
+
+        visitor = visitor.set_spec_version(spec_version.clone())?;
+
+        let attributes = spec_version.attribute_names();
+
+        for (hn, hv) in self
+            .headers
+            .iter()
+            .filter(|(hn, _)| "ce_specversion" != **hn && hn.starts_with("ce_"))
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:80]: Can you use the constant `SPEC_VERSION_HEADER`?

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs (hunk 5 above)
Comment (slinkydeveloper) [reply cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:14]: Ok now that i look at it, it doesn't really make much difference because then you buffer everything in order to go to event... I wonder, can you buffer to `Vec<u8>` more than using the `Bytes` struct? `Bytes` is meant to be a flexible byte buffer, we don't need its functionalities here

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        {
+            let name = &hn["ce_".len()..];
+
+            if attributes.contains(&name) {
+                visitor = visitor.set_attribute(
+                    name,
+                    MessageAttributeValue::String(String::from(header_value_to_str!(hv)?)),
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:87]: If you change from the comments above the header value from `Bytes` to `Vec<u8>`, then you don't need this dance here to convert to string. You just need `String::from_utf8` (which also avoids another unnecessary copy you're doing here)

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+            return Err(message::Error::WrongEncoding {});
+        }
+        visitor.set_structured_event(self.payload.to_vec())
+    }
+}
+
+impl MessageDeserializer for ConsumerRecordDeserializer {
+    fn encoding(&self) -> Encoding {
+        match (
+            str::from_utf8(self.headers.get("content-type").unwrap_or(&Bytes::new()))
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:124]: This condition can be simplified with a `map` step. Given the change about `Vec<u8>`, you can do something like `self.headers.get("content-type").map(|s| String::from_utf8(s).ok()).flatten().map(|s| s.starts_with(...))`

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+                .unwrap_or("UNKNOWN")
+                .starts_with("application/cloudevents+json"),
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:126]: You have the constant for that :)

File: cloudevents-sdk-rdkafka/src/kafka_producer_record.rs [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+/// Method to fill a [`RequestBuilder`] with an [`Event`]
+pub fn event_to_record<'a, K: ToBytes + ?Sized>(
+    event: &'a ProducerRecordSerializer,
+    mut record: FutureRecord<'a, K, Vec<u8>>,
+) -> Result<FutureRecord<'a, K, Vec<u8>>> {
+    let header = event.headers.clone();
+
+    record = record.headers(header);
+    record = record.payload(event.payload.as_ref().unwrap());
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:80]: You should add the payload only if there is any

File: cloudevents-sdk-rdkafka/src/kafka_producer_record.rs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+impl BinarySerializer<ProducerRecordSerializer> for ProducerRecordSerializer {
+    fn set_spec_version(mut self, spec_version: SpecVersion) -> Result<Self> {
+        self.headers = self.headers.add(SPEC_VERSION_HEADER, spec_version.as_str());
+
+        Ok(self)
+    }
+
+    fn set_attribute(mut self, name: &str, value: MessageAttributeValue) -> Result<Self> {
+        self.headers = self.headers.add(
+            &headers::ATTRIBUTES_TO_HEADERS.get(name).unwrap().clone()[..],
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:36]: Why this unwrap? what if the user provides an invalid `name`?

File: cloudevents-sdk-rdkafka/src/headers.rs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    ($header_value:expr) => {
+        str::from_utf8($header_value).map_err(|e| cloudevents::message::Error::Other {
+            source: Box::new(e),
+        })
+    };
+}
+
+macro_rules! attribute_name_to_header {
+    ($attribute:expr) => {
+        format!("{}{}", "ce_", $attribute)
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/headers.rs:15]: You can simplify to `format!("ce_{}", $attribute)`

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs (hunk 5 above)
Comment (pranav-bhatt) [reply cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:14]: Yeah on it. I put bytes cause i didn't really know at the time if i would need to do said fancy stuff, and then i completely forgot about it.

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs (hunk 7 above)
Comment (pranav-bhatt) [reply cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:87]: No I made a mistake here. I changed the function in headers from `str::from_utf8` to `String::from_utf8`. No more unnecessary copy :)

File: cloudevents-sdk-rdkafka/src/kafka_producer_record.rs (hunk 11 above)
Comment (pranav-bhatt) [reply cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:36]: So what should i return if the name is wrong?

Comment (slinkydeveloper) [reply cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:36]: i suppose it should return some err? Does `message::Error` already have a variant like "invalid attribute name"?

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs (hunk 7 above)
Comment (slinkydeveloper) [reply cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:87]: No you don't, because `hv` should be already a `Vec<u8>` with the proposed change above

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+        for i in 0..headers.count() {
+            let header = headers.get(i).unwrap();
+            resp_des
+                .headers
+                .insert(header.0.to_string(), Vec::from(header.1));
+        }
+
+        match message.payload() {
+            Some(s) => resp_des.payload = Vec::from(s),
+            None => resp_des.payload = resp_des.payload,
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:34]: Do you really need this match arm?

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:55]: Same comment as above

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+    pub(crate) payload: Vec<u8>,
+}
+
+impl ConsumerRecordDeserializer {
+    pub fn owned_new(message: OwnedMessage) -> ConsumerRecordDeserializer {
+        let mut resp_des = ConsumerRecordDeserializer {
+            headers: HashMap::new(),
+            payload: Vec::new(),
+        };
+        let headers = message.headers().unwrap();
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:24]: is this `unwrap()` safe? should we return an error in case no headers are there?

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+
+        resp_des
+    }
+
+    pub fn borrowed_new(message: &BorrowedMessage) -> ConsumerRecordDeserializer {
+        let mut resp_des = ConsumerRecordDeserializer {
+            headers: HashMap::new(),
+            payload: Vec::new(),
+        };
+        let headers = message.headers().unwrap();
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:45]: same comment as above

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+impl ConsumerRecordDeserializer {
+    pub fn owned_new(message: OwnedMessage) -> ConsumerRecordDeserializer {
+        let mut resp_des = ConsumerRecordDeserializer {
+            headers: HashMap::new(),
+            payload: Vec::new(),
+        };
+        let headers = message.headers().unwrap();
+        for i in 0..headers.count() {
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:25]: Can you make a function to use that loop? notice that both `OwnedHeaders` and `BorrowedHeaders` implements `Headers` trait, so you can make a function that gets `impl Headers` and return the hashmap of headers

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+};
+use cloudevents::{message, Event};
+use rdkafka::message::{BorrowedMessage, Headers, Message, OwnedMessage};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::str;
+
+pub struct ConsumerRecordDeserializer {
+    pub(crate) headers: HashMap<String, Vec<u8>>,
+    pub(crate) payload: Vec<u8>,
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:15]: this should be `Option<Vec<u8>>` imo

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+    use url::Url;
+
+    #[test]
+    fn test_binary_record() {
+        let time = Utc::now();
+
+        let expected = EventBuilderV10::new()
+            .id("0001")
+            .ty("example.test")
+            //TODO this is required now because the message deserializer implictly set default values
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:201]: that doesn't happen anymore, you can remove this comment

File: cloudevents-sdk-rdkafka/src/kafka_producer_record.rs [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+};
+use cloudevents::Event;
+use rdkafka::message::{OwnedHeaders, ToBytes};
+use rdkafka::producer::FutureRecord;
+
+/// Wrapper for [`RequestBuilder`] that implements [`StructuredSerializer`] & [`BinarySerializer`] traits
+#[derive(Debug)]
+pub struct ProducerRecordSerializer {
+    pub(crate) payload: Option<Vec<u8>>,
+    pub(crate) headers: OwnedHeaders,
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:14]: headers goes before payload

File: cloudevents-sdk-rdkafka/src/kafka_producer_record.rs [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+use cloudevents::event::SpecVersion;
+use cloudevents::message::{
+    BinaryDeserializer, BinarySerializer, MessageAttributeValue, Result, StructuredSerializer,
+};
+use cloudevents::Event;
+use rdkafka::message::{OwnedHeaders, ToBytes};
+use rdkafka::producer::FutureRecord;
+
+/// Wrapper for [`RequestBuilder`] that implements [`StructuredSerializer`] & [`BinarySerializer`] traits
+#[derive(Debug)]
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:11]: Any reason you need derive `Debug`? can we remove it?

File: cloudevents-sdk-rdkafka/src/kafka_producer_record.rs [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    fn set_attribute(mut self, name: &str, value: MessageAttributeValue) -> Result<Self> {
+        self.headers = self.headers.add(
+            &headers::ATTRIBUTES_TO_HEADERS
+                .get(name)
+                .ok_or(cloudevents::message::Error::UnrecognizedAttributeName {
+                    name: String::from(name),
+                })
+                .unwrap()
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:42]: why unwrap? you should use the `?` to propagate the error

File: example-projects/rdkafka-example/src/main.rs [hunk 23]
```diff
@@ -0,0 +1,249 @@
...
+        })
+        .collect::<Vec<_>>();
+
+    // This loop will wait until all delivery statuses have been received.
+    for future in futures {
+        info!("Future completed. Result: {:?}", future.await);
+    }
+}
+
+#[tokio::main]
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:174]: should tokio main macro annotate main function?

File: example-projects/rdkafka-example/src/main.rs [hunk 24]
```diff
@@ -0,0 +1,249 @@
...
+
+                consumer.commit_message(&m, CommitMode::Async).unwrap();
+            }
+        };
+    }
+}
+
+#[tokio::main]
+async fn consumer_example(log: Option<&str>, brokers: &str, group_id: &str, topics: Vec<&str>) {
+    setup_logger(true, log);
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:120]: can we remove all this logger noise? it just makes the example more convoluted

File: example-projects/rdkafka-example/src/main.rs [hunk 25]
```diff
@@ -0,0 +1,249 @@
...
+    // consumer.start() returns a stream. The stream can be used ot chain together expensive steps,
+    // such as complex computations on a thread pool or asynchronous IO.
+    let mut message_stream = consumer.start();
+
+    while let Some(message) = message_stream.next().await {
+        match message {
+            Err(e) => warn!("Kafka error: {}", e),
+            Ok(m) => {
+                let event = m.into_event().unwrap();
+                info!("Received Event: {:#?}", event);
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:110]: just use println!, without the logging macro

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs (hunk 13 above)
Comment (pranav-bhatt) [reply cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:34]: this is one of those "ill deal with later" things i was talking about. Has been sorted

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs (hunk 15 above)
Comment (pranav-bhatt) [reply cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:24]: yeah it seems headers are more of a feature in kafka than a necessity.

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs (hunk 17 above)
Comment (pranav-bhatt) [reply cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:25]: I tried that a while back and it didnt work cause the type `&BorrowedMessage` doesn't implement `Headers`

File: example-projects/rdkafka-example/src/main.rs (hunk 23 above)
Comment (pranav-bhatt) [reply example-projects/rdkafka-example/src/main.rs:174]: silly copy paste mistake :P

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs (hunk 17 above)
Comment (pranav-bhatt) [reply cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:25]: ok so now im taking reference for both, and am able to do something similar

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 26]
```diff
@@ -0,0 +1,253 @@
...
+    }
+}
+
+/// Method to transform an incoming [`Response`] to [`Event`]
+pub fn record_to_event(msg: &impl Message) -> Result<Event> {
+    MessageDeserializer::into_event(ConsumerRecordDeserializer::new(msg)?)
+}
+
+pub trait BorrowedMessageExt {
+    fn event_from(&self) -> Result<Event>;
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:137]: What about `to_event()`?

File: cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs [hunk 27]
```diff
@@ -0,0 +1,253 @@
...
+}
+
+impl BorrowedMessageExt for BorrowedMessage<'_> {
+    fn event_from(&self) -> Result<Event> {
+        record_to_event(self)
+    }
+}
+
+pub trait OwnedMessageExt {
+    fn event_from(&self) -> Result<Event>;
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_consumer_record.rs:147]: Same as above

File: cloudevents-sdk-rdkafka/src/kafka_producer_record.rs [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+/// Method to fill a [`RequestBuilder`] with an [`Event`]
+pub fn event_to_record<'a, K: ToBytes + ?Sized>(
+    event: &'a ProducerRecordSerializer,
+    mut record: FutureRecord<'a, K, Vec<u8>>,
+) -> Result<FutureRecord<'a, K, Vec<u8>>> {
+    let header = event.headers.clone();
+
+    record = record.headers(header);
+
+    match event.payload.as_ref() {
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:88]: What about replacing this match statement with `if let`?

File: cloudevents-sdk-rdkafka/src/kafka_producer_record.rs [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+impl<'a, K: ToBytes + ?Sized> FutureRecordExt<'a, K> for FutureRecord<'a, K, Vec<u8>> {
+    fn event(self, event: &'a ProducerRecordSerializer) -> Result<FutureRecord<'a, K, Vec<u8>>> {
+        event_to_record(event, self)
+    }
+}
+
+pub trait EventExt {
+    fn serialize_event(self) -> Result<ProducerRecordSerializer>;
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/src/kafka_producer_record.rs:108]: I don't understand this trait, should the user just use the previous one? and why the previous one accepts `ProducerRecordSerializer` more than `Event`?

File: example-projects/rdkafka-example/src/main.rs [hunk 30]
```diff
@@ -0,0 +1,187 @@
...
+
+use rdkafka::client::ClientContext;
+use rdkafka::config::{ClientConfig, RDKafkaLogLevel};
+use rdkafka::consumer::stream_consumer::StreamConsumer;
+use rdkafka::consumer::{CommitMode, Consumer, ConsumerContext, Rebalance};
+use rdkafka::error::KafkaResult;
+use rdkafka::producer::{FutureProducer, FutureRecord};
+use rdkafka::topic_partition_list::TopicPartitionList;
+
+// run a kafka lense: docker run --rm --net=host -e ADV_HOST=localhost -e SAMPLEDATA=0 lensesio/fast-data-dev
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:18]: s/run a kafka lense/run a Kafka Docker container, for example

File: example-projects/rdkafka-example/src/main.rs [hunk 31]
```diff
@@ -0,0 +1,187 @@
...
+        println!("Post rebalance {:?}", rebalance);
+    }
+
+    fn commit_callback(&self, result: KafkaResult<()>, _offsets: &TopicPartitionList) {
+        println!("Committing offsets: {:?}", result);
+    }
+}
+
+// A type alias with your custom consumer can be created for convenience.
+type LoggingConsumer = StreamConsumer<CustomContext>;
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:42]: Can we remove this type alias?

File: example-projects/rdkafka-example/src/main.rs [hunk 32]
```diff
@@ -0,0 +1,187 @@
...
+
+async fn consume_and_print(brokers: &str, group_id: &str, topics: &[&str]) {
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:44]: just `consume` is fine

File: example-projects/rdkafka-example/src/main.rs [hunk 33]
```diff
@@ -0,0 +1,187 @@
...
+    // This loop is non blocking: all messages will be sent one after the other, without waiting
+    // for the results.
+    let futures = (0..5)
+        .map(|i| async move {
+            // The send operation on the topic returns a future, which will be
+            // completed once the result or failure from Kafka is received.
+            let event = EventBuilderV10::new()
+                .id("0001")
+                .ty("example.test")
+                .source(Url::from_str("http://localhost/").unwrap())
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:96]: you don't need that, you can just do `source("http://localhost/")`

File: example-projects/rdkafka-example/src/main.rs [hunk 34]
```diff
@@ -0,0 +1,187 @@
...
+        .expect("Producer creation error");
+
+    // This loop is non blocking: all messages will be sent one after the other, without waiting
+    // for the results.
+    let futures = (0..5)
+        .map(|i| async move {
+            // The send operation on the topic returns a future, which will be
+            // completed once the result or failure from Kafka is received.
+            let event = EventBuilderV10::new()
+                .id("0001")
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:94]: what about using `i` as the id?

File: example-projects/rdkafka-example/src/main.rs [hunk 35]
```diff
@@ -0,0 +1,187 @@
...
+                .ty("example.test")
+                .source(Url::from_str("http://localhost/").unwrap())
+                .data("application/json", json!({"hello": "world"}))
+                .extension("someint", "10")
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:98]: Remove this, it's noise

File: example-projects/rdkafka-example/src/main.rs [hunk 36]
```diff
@@ -0,0 +1,187 @@
...
+
+    // This loop will wait until all delivery statuses have been received.
+    for future in futures {
+        println!("Future completed. Result: {:?}", future.await);
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let selector = App::new("Main Application")
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:128]: Main application? what about `CloudEvents Kafka Example`?

File: example-projects/rdkafka-example/src/main.rs [hunk 37]
```diff
@@ -0,0 +1,183 @@
...
+use rdkafka::error::KafkaResult;
+use rdkafka::producer::{FutureProducer, FutureRecord};
+use rdkafka::topic_partition_list::TopicPartitionList;
+
+// run a Kafka lense or a Kafka Docker container to try out this example
+
+// A context can be used to change the behavior of producers and consumers by adding callbacks
+// that will be executed by librdkafka.
+// This particular context sets up custom callbacks to log rebalancing events.
+struct CustomContext;
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:23]: Remove this, you can use the default consumer context https://fede1024.github.io/rust-rdkafka/rdkafka/consumer/struct.DefaultConsumerContext.html

File: cloudevents-sdk-rdkafka/README.md [hunk 38]
```diff
@@ -0,0 +1,28 @@
+# CloudEvents SDK Rust - Reqwest [![Crates badge]][crates.io] [![Docs badge]][docs.rs]
```
Comment (slinkydeveloper) [inline cloudevents-sdk-rdkafka/README.md:1]: s/Reqwest/rdkafka

File: example-projects/rdkafka-example/Cargo.toml [hunk 39]
```diff
@@ -0,0 +1,25 @@
...
+serde_json = "^1.0"
+futures = "^0.3"
+tokio = { version = "^0.2", features = ["full"] }
+clap = "2.33.1"
+
+[dependencies.rdkafka]
+version = "~0.23"
+features = ["ssl", "sasl"]
+
+[workspace]
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/Cargo.toml:25]: workspace? you don't need this

File: example-projects/rdkafka-example/src/main.rs [hunk 40]
```diff
@@ -0,0 +1,160 @@
...
+
+use cloudevents::{EventBuilder, EventBuilderV10};
+use cloudevents_sdk_rdkafka::{FutureRecordExt, MessageExt, MessageRecord};
+
+use rdkafka::config::{ClientConfig, RDKafkaLogLevel};
+use rdkafka::consumer::stream_consumer::StreamConsumer;
+use rdkafka::consumer::{CommitMode, Consumer, DefaultConsumerContext};
+use rdkafka::producer::{FutureProducer, FutureRecord};
+
+// run a Kafka lense or a Kafka Docker container to try out this example
```
Comment (slinkydeveloper) [inline example-projects/rdkafka-example/src/main.rs:13]: Why did you removed the docker command? it was a good one, just the name is not kafka lense :smile: 

Maybe rephrase with:

> You need a running Kafka cluster to try out this example. With docker: [insert command her
